### PR TITLE
Log reg fit was patched

### PR DESCRIPTION
**Pull request by masdevas**
Log reg fit was patched

### CODE REVIEW

File: daal4py/sklearn/linear_model/_logistic_path_0_22.py [hunk 1]
```diff
@@ -1012,29 +1009,176 @@ def daal4py_predict(self, X, resultsToEvaluate):
     raise ValueError('resultsToEvaluate must be in [computeClassLabels, \
                      computeClassProbabilities, computeClassLogProbabilities]')
 
+def daal4py_fit(self, X, y, sample_weight=None):
```
Comment (PetrovKP) [inline daal4py/sklearn/linear_model/_logistic_path_0_22.py:1012]: does it make sense to rename the file?

Comment (PetrovKP) [inline daal4py/sklearn/linear_model/_logistic_path_0_22.py:1012]: not critical, but it is better to name it with the prefix daal4py, what will call directly daal4py alforithms

File: daal4py/sklearn/linear_model/_logistic_path_0_22.py [hunk 2]
```diff
@@ -1012,29 +1009,176 @@ def daal4py_predict(self, X, resultsToEvaluate):
...
+    self : object
+    """
+    solver = _check_solver(self.solver, self.penalty, self.dual)
+    daal_ready = solver in ['lbfgs', 'newton-cg'] and not sparse.issparse(X)
+    daal_ready = daal_ready and sample_weight is None and self.class_weight is None
+
+    if daal_ready:
+        logging.info("sklearn.linear_model.LogisticRegression.fit: " + get_patch_message("daal"))
+    else:
+        logging.info("sklearn.linear_model.LogisticRegression.fit: " + get_patch_message("sklearn"))
```
Comment (PetrovKP) [inline daal4py/sklearn/linear_model/_logistic_path_0_22.py:1039]: in this place we can call the base class (original function) and not copy a lot of code next. right?

File: daal4py/sklearn/linear_model/_logistic_path_0_22.py [hunk 3]
```diff
@@ -1012,29 +1009,176 @@ def daal4py_predict(self, X, resultsToEvaluate):
...
+
```
Comment (PetrovKP) [inline daal4py/sklearn/linear_model/_logistic_path_0_22.py:1040]: I think we are unnecessarily copying so much code. Leave code branches that are only needed for call of daal4py

File: daal4py/sklearn/linear_model/_logistic_path_0_22.py [hunk 4]
```diff
@@ -21,12 +21,12 @@
 import numbers
```
Comment (PetrovKP) [inline daal4py/sklearn/linear_model/_logistic_path_0_22.py:21]: 1200+ lines! I think it makes sense to split the file. Good practice

File: daal4py/sklearn/monkeypatch/dispatcher.py [hunk 5]
```diff
@@ -96,7 +88,9 @@ def _get_map_of_algorithms():
         'fin_check':                [[(validation, '_assert_all_finite', _daal_assert_all_finite), None]],
         'tsne':                     [[(manifold_module, 'TSNE', TSNE_daal4py), None]],
     }
-    if daal_check_version((2021,'P', 100)):
+    if LooseVersion(sklearn_version) < LooseVersion("0.22"):
+        mapping['logistic'] = [[(logistic_module, _patched_log_reg_path_func_name, daal_optimized_logistic_path), None]]
+    else:
         mapping['log_reg'] = [[(linear_model_module, 'LogisticRegression', LogisticRegression_daal4py), None]]
```
Comment (PetrovKP) [inline daal4py/sklearn/monkeypatch/dispatcher.py:94]:
```suggestion
        mapping['logistic'] = [[(linear_model_module, 'LogisticRegression', LogisticRegression_daal4py), None]]
```

File: daal4py/sklearn/monkeypatch/dispatcher.py [hunk 6]
```diff
@@ -96,7 +88,9 @@ def _get_map_of_algorithms():
         'fin_check':                [[(validation, '_assert_all_finite', _daal_assert_all_finite), None]],
         'tsne':                     [[(manifold_module, 'TSNE', TSNE_daal4py), None]],
     }
-    if daal_check_version((2021,'P', 100)):
+    if LooseVersion(sklearn_version) < LooseVersion("0.22"):
```
Comment (PetrovKP) [inline daal4py/sklearn/monkeypatch/dispatcher.py:91]: Does it make sense to support version 0.21? It's very hard?

File: daal4py/sklearn/utils/fixes.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
+from functools import update_wrapper
```
Comment (PetrovKP) [inline daal4py/sklearn/utils/fixes.py:1]: Copyright + I don't really understand why you need this file and functions? Why not take these functions from Scikit-learn?

**Review by PetrovKP — COMMENTED** [review]
You have a lot of falls in public CI. You should green public CI

**Comment by PetrovKP** [comment]
@Pahandrovich , Please see this PR for security issues.

File: daal4py/sklearn/linear_model/_logistic_path_0_22.py (hunk 1 above)
Comment (masdevas) [reply daal4py/sklearn/linear_model/_logistic_path_0_22.py:1012]: daal4py_fit is called only directly from LogisticRegression class

File: daal4py/sklearn/monkeypatch/dispatcher.py (hunk 6 above)
Comment (masdevas) [reply daal4py/sklearn/monkeypatch/dispatcher.py:89]: I don't think that performance improve for old versions is critical for us

Comment (PetrovKP) [reply daal4py/sklearn/monkeypatch/dispatcher.py:89]: If merging the files 0.21 and 0.22 is not difficult, I would do it.

File: daal4py/sklearn/linear_model/logistic_loss.py [hunk 8]
```diff
@@ -88,7 +88,7 @@ def _daal4py_loss_and_grad(beta, objF_instance, X, y, n):
     beta_ = make2d(beta)
     res = objF_instance.compute(X, y, beta_)
     gr = res.gradientIdx.ravel()
-    gr = gr * n # force copy
+    gr = gr.astype(np.float64) * n # force copy
```
Comment (PetrovKP) [inline daal4py/sklearn/linear_model/logistic_loss.py:91]: if gr is double type - will copying take place here?

File: daal4py/sklearn/linear_model/_logistic_0_22.py [hunk 9]
```diff
@@ -1012,29 +895,132 @@ def daal4py_predict(self, X, resultsToEvaluate):
...
+    """
+    solver = _check_solver(self.solver, self.penalty, self.dual)
+    daal_ready = solver in ['lbfgs', 'newton-cg'] and not sparse.issparse(X)
+    daal_ready = daal_ready and sample_weight is None and self.class_weight is None
+
+    if daal_ready:
+        logging.info("sklearn.linear_model.LogisticRegression.fit: " + get_patch_message("daal"))
+    else:
+        logging.info("sklearn.linear_model.LogisticRegression.fit: " + get_patch_message("sklearn"))
+        return LogisticRegression_original.fit(self, X, y, sample_weight=sample_weight)
```
Comment (PetrovKP) [inline daal4py/sklearn/linear_model/_logistic_0_22.py:926]: from 918-926 correctly (see function name) transfer to method fit

**Pull request closed without merging**